### PR TITLE
[9.0][IMP] stock_cycle_count: Add views and reports

### PR DESCRIPTION
**Pull request by LoisRForgeFlow**
Improve tools to analyse how accuracy evolve in locations.

### CODE REVIEW

File: stock_cycle_count/models/stock_inventory.py [hunk 1]
```diff
@@ -12,23 +12,27 @@
...
+        for inv in self:
+            theoretical = sum(inv.line_ids.mapped(
+                lambda x: abs(x.theoretical_qty)))
+            abs_discrepancy = sum(inv.line_ids.mapped(
+                lambda x: abs(x.discrepancy_qty)))
+            if theoretical:
+                inv.inventory_accuracy = max(
+                    PERCENT * (theoretical - abs_discrepancy) / theoretical,
+                    0.0)
+            if not inv.line_ids and inv.state == 'done':
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_inventory.py:27]: You can get these records out of the loop (as you based your computation on inv_lines : substract self recordset with records with no inv_lines)

File: stock_cycle_count/reports/report_stock_location_accuracy.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+from openerp import api, fields, models
+
+
+class LocationAccuracyReport(models.AbstractModel):
+    _name = "report.stock_location_accuracy"
+
+    @api.model
+    def _get_inventory_domain(self, loc_id):
+        return [('location_id', '=', loc_id),
+                ('exclude_sublocation', '=', True),
```
Comment (rousseldenis) [inline stock_cycle_count/reports/report_stock_location_accuracy.py]: Is this shouldn't be optional through parameter ?

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review. Minor comment.

Good improvement!

Same remarks about api on v10 (to make travis green)

File: stock_cycle_count/models/stock_inventory.py (hunk 1 above)
Comment (LoisRForgeFlow) [reply stock_cycle_count/models/stock_inventory.py:27]: if I filter the recordset as you propose the inventories with no lines and in state done will end up with an accuracy of 0.0 (the default), while here I'm setting them to 100.0. Am I missing something?

File: stock_cycle_count/reports/report_stock_location_accuracy.py (hunk 2 above)
Comment (LoisRForgeFlow) [reply stock_cycle_count/reports/report_stock_location_accuracy.py]: yes, why not? adding now

File: stock_cycle_count/models/stock_inventory.py (hunk 1 above)
Comment (rousseldenis) [reply stock_cycle_count/models/stock_inventory.py:27]: No, I want to say you should compute this in batch before the loop and substract these recordset to enter the loop

Comment (LoisRForgeFlow) [reply stock_cycle_count/models/stock_inventory.py:27]: Ok, I got you. Fixed in https://github.com/OCA/stock-logistics-warehouse/pull/346/commits/d1832362eddedebd13813202aa4709f768991611

**Comment by rousseldenis** [comment]
@lreficent Can you check travis ?

File: stock_cycle_count/models/stock_inventory.py [hunk 3]
```diff
@@ -15,7 +15,10 @@ class StockInventory(models.Model):
     @api.multi
     @api.depends("state", "line_ids")
     def _compute_inventory_accuracy(self):
-        for inv in self:
+        zero_inventories = self.filtered(
+            lambda i: not i.line_ids and i.state == 'done')
+        zero_inventories.inventory_accuracy = PERCENT
```
Comment (LoisRForgeFlow) [inline stock_cycle_count/models/stock_inventory.py]: @rousseldenis This line of course give singleton error, but the obivous solution (`zero_inventories.write({'inventory_accuracy': PERCENT})`) is creating an infinite loop, I really don't understand why. Any idea?

File: stock_cycle_count/models/stock_inventory.py [hunk 4]
```diff
@@ -17,7 +17,8 @@ class StockInventory(models.Model):
     def _compute_inventory_accuracy(self):
         zero_inventories = self.filtered(
             lambda i: not i.line_ids and i.state == 'done')
-        zero_inventories.inventory_accuracy = PERCENT
+        for inv in zero_inventories:
+            inv.inventory_accuracy = PERCENT
```
Comment (LoisRForgeFlow) [inline stock_cycle_count/models/stock_inventory.py]: And the only way to fix it is this commit (https://github.com/OCA/stock-logistics-warehouse/pull/346/commits/2b4103365fa4c31e041fda3f8ee387c1aa967e97) which I think leave us in a position very similar to the starting one and I hardly see any performance improvement. What do you think?

**Comment by LoisRForgeFlow** [comment]
@rousseldenis See my two comments above.

File: stock_cycle_count/models/stock_inventory.py (hunk 4 above)
Comment (rousseldenis) [reply stock_cycle_count/models/stock_inventory.py]: No, in fact. I didn't realized.

Indeed you cannot write on compute. Wrong! You have to change recordset fields one by one(the save is done at the end).

The improvement is only valid outside computes. Sorry.

Comment (LoisRForgeFlow) [reply stock_cycle_count/models/stock_inventory.py]: I didn't get a notification of this answer. No worries, it is always good to learn :wink:

Comment (LoisRForgeFlow) [reply stock_cycle_count/models/stock_inventory.py]: @rousseldenis reverted to initial state, could you please review again?

**Review by rousseldenis — APPROVED** [review]

**Comment by aheficent** [comment]
:heart_eyes:

**Review by MiquelRForgeFlow — APPROVED** [review]

**Pull request merged**